tests/conftest.py

```diff
@@ -124,57 +125,100 @@ def clean_serveradmin(filters):
+                    str(AWS_RETURN_CODES['pending']),
+                    str(AWS_RETURN_CODES['running']),
+                    str(AWS_RETURN_CODES['shutting-down']),
+                    str(AWS_RETURN_CODES['terminated']),
+                    str(AWS_RETURN_CODES['stopping']),
+                    str(AWS_RETURN_CODES['stopped']),
+                ]
+            },
+        ],
+        InstanceIds=[instance_id],
+        DryRun=False)['Reservations'][0]['Instances'][0]['State']['Code']
+
+    return int(response)
 
+
+def clean_aws(vm_hostname):
     try:
         obj = Query({'hostname': vm_hostname}, ['aws_instance_id']).get()
     except DatasetError:  # No object to clean up
         return
 
+    timeout = 120
+    ec2 = boto3.client('ec2')
     if not obj['aws_instance_id']:
         return
 
-    timeout = 120
-    ec2 = boto3.client('ec2')
     try:
         ec2.stop_instances(
             InstanceIds=[obj['aws_instance_id']], DryRun=False
         )
+        for _ in range(timeout):
+            instance_status = _get_aws_instance_status(obj['aws_instance_id'])
+            if AWS_RETURN_CODES['stopped'] == instance_status:
+                break
+            sleep(1)
     except ClientError as e:
         pass  # Not running
-    for _ in range(timeout):
-        instance_status = _get_instance_status()
-        if AWS_RETURN_CODES['stopped'] == instance_status:
-            break
-        sleep(1)
 
     ec2.terminate_instances(InstanceIds=[obj['aws_instance_id']])
     for _ in range(timeout):
-        instance_status = _get_instance_status()
+        instance_status = _get_aws_instance_status(obj['aws_instance_id'])
         if AWS_RETURN_CODES['terminated'] == instance_status:
             break
         sleep(1)
 
 
+def clean_orphaned_aws_vms():
+    def _get_leftover_vms() -> List:
+        try:
+            response = ec2.describe_instances(
+                Filters=[
+                    {
+                        'Name': 'tag:Name',
+                        'Values': ['?igvm*']
+                    },
+                    {
+                        'Name': 'instance-state-code',
+                        'Values': [
+                            str(AWS_RETURN_CODES['pending']),
+                            str(AWS_RETURN_CODES['running']),
+                            str(AWS_RETURN_CODES['shutting-down']),
+                            str(AWS_RETURN_CODES['stopping']),
+                            str(AWS_RETURN_CODES['stopped']),
+                        ]
+                    },
+                ]
+            )['Reservations']
+
+            instance_ids = [r['Instances'][0]['InstanceId'] for r in response]
+        except IndexError:
+            return []
+
+        return instance_ids
+
+    timeout = 120
+    ec2 = boto3.client('ec2')
+    leftover_vm_ids = _get_leftover_vms()
+    if len(leftover_vm_ids) > 0:
+        ec2.terminate_instances(InstanceIds=leftover_vm_ids)
+        for _ in range(timeout):
```

Comment:
i counted this kind of loop now 3 times duplicated in this diff. lets move this out into a shared function

Reply:
moved to a shared function called `_wait_for_state_reached`